File: src/content/docs/topics/cs/coding-problems/binary-search/004-median-of-two-sorted-arrays-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
    const m = nums1.length, n = nums2.length;
    const total = m + n;
    const need = Math.floor(total / 2) + 1;

    let i = 0, j = 0;
    let prev = 0, cur = 0;
    for (let step = 0; step < need; step++) {
        prev = cur;
        if (i < m && (j >= n || nums1[i] <= nums2[j])) {
            cur = nums1[i++];
        } else {
            cur = nums2[j++];
        }
    }
    return total % 2 === 1 ? cur : (prev + cur) / 2;
}

assert(findMedianSortedArrays([1, 3], [2]) === 2.0);
assert(findMedianSortedArrays([1, 2], [3, 4]) === 2.5);
assert(findMedianSortedArrays([0, 0], [0, 0]) === 0.0);
assert(findMedianSortedArrays([], [1]) === 1.0);
assert(findMedianSortedArrays([2], []) === 2.0);
assert(findMedianSortedArrays([1, 3], [2, 4]) === 2.5);
console.log("all tests pass");

File: src/content/docs/topics/cs/coding-problems/binary-search/004-median-of-two-sorted-arrays-approach2.py
def find_median_sorted_arrays(nums1: list, nums2: list) -> float:
    m, n = len(nums1), len(nums2)
    total = m + n
    need = total // 2 + 1

    i = j = 0
    prev = cur = 0
    for _ in range(need):
        prev = cur
        if i < m and (j >= n or nums1[i] <= nums2[j]):
            cur = nums1[i]
            i += 1
        else:
            cur = nums2[j]
            j += 1

    return float(cur) if total % 2 == 1 else (prev + cur) / 2

assert find_median_sorted_arrays([1, 3], [2]) == 2.0
assert find_median_sorted_arrays([1, 2], [3, 4]) == 2.5
assert find_median_sorted_arrays([0, 0], [0, 0]) == 0.0
assert find_median_sorted_arrays([], [1]) == 1.0
assert find_median_sorted_arrays([2], []) == 2.0
assert find_median_sorted_arrays([1, 3], [2, 4]) == 2.5
print("all tests pass")
